File: scripts/perfgate.py
#!/usr/bin/env python3
"""
Refuse a commit that made the frame slower (5.7).

    scripts/perfgate.py                  check the tree against perf-budget.json
    scripts/perfgate.py --update         re-baseline, deliberately
    scripts/perfgate.py --config debug   check a different configuration

Exit 0 inside budget, 1 over it, 2 if it could not measure. Installed as a `pre-commit`
hook by `scripts/install_hooks.sh`, where it runs **only when a commit touches something
that can move the frame** -- `engine/gfx/`, `engine/shaders/` or `engine/scene/`. A doc
commit pays nothing, which is what keeps the hook from being the thing everyone passes
`--no-verify` to.

## It gates two zones, and the other four would make it lie

`Lighting` and `Frame` only. `GBuffer`, `SSAO`, `SSR` and `Bloom` settle into one of two
whole-run states about 5% apart -- tooling.md, "The bimodal zones" -- so a gate on them
fails on a coin flip, and a gate that fails at random is one nobody reads. `Lighting` and
`Frame` are the two that hold, which is why the reference quotes those and not the others.

## The minimum, not the median, and that is the whole trick

A pre-commit hook runs on a machine that is also doing something else: another session's
build, a browser, the shader compile that just finished. Load moves a median a long way and
a minimum barely at all -- a loaded run still contains frames that ran unobstructed. Taking
the per-frame minimum is what makes this measure the *code* rather than the machine, and it
is why the budget can be tight enough to catch something real.

The margin is **12%** over the recorded minimum. That is far outside the 5% bimodality the
reference documents and far inside the kind of regression worth blocking a commit for -- the
incident this was written after was about 2x. A change that genuinely costs 12% is one that
should be re-baselined on purpose rather than absorbed.

## Re-baselining is deliberate, exactly as a golden snap is

`--update` rewrites `perf-budget.json`, which is committed, so the new number arrives in a
diff somebody reads. There is no automatic drift: a budget that quietly followed the tree
upward would report green while the frame got slower every week, which is the failure this
exists to prevent.
"""

import argparse
import json
import pathlib
import statistics
import subprocess
import sys
import tempfile

ROOT = pathlib.Path(__file__).resolve().parent.parent
BUDGET = ROOT / "perf-budget.json"

# The arm. Fixed on purpose: a gate whose scene or camera moved with the tree would compare
# two different pictures and call the difference a regression. `--locked` pins the clock,
# `--headless` never maps a window (so a hook cannot steal focus), and the camera is written
# out rather than left to `frameBounds`, which follows the scene's bounds and would shift the
# moment an asset did.
ARM = ["--locked", "--audio-null", "--headless", "--msaa", "4",
       "--camera", "0.00,1.16,1.80,63.6,-12.6,7.87"]
FRAMES = 300
ZONES = ("Lighting", "Frame")
MARGIN = 1.12


def measure(config: str) -> dict:
    """Per-frame minimum of each gated zone, in ms. Returns {} if nothing was traced."""
    binary = ROOT / "build" / config / "demo"
    if not binary.exists():
        print(f"perfgate: {binary} is not built. Run: ./build_game.sh demo {config}",
              file=sys.stderr)
        sys.exit(2)

    with tempfile.TemporaryDirectory() as tmp:
        trace = pathlib.Path(tmp) / "trace.json"
        cmd = ["timeout", "-s", "TERM", "180", "./run.sh", config, "--",
               *ARM, "--frames", str(FRAMES), "--trace", str(trace)]
        proc = subprocess.run(cmd, cwd=ROOT, capture_output=True, text=True)
        if not trace.exists():
            print(f"perfgate: no trace written (exit {proc.returncode}); not measured",
                  file=sys.stderr)
            print(proc.stderr[-2000:], file=sys.stderr)
            sys.exit(2)
        events = json.loads(trace.read_text())
        events = events["traceEvents"] if isinstance(events, dict) else events

    per_zone: dict[str, list[float]] = {}
    for e in events:
        if e.get("ph") == "X" and e.get("cat") == "gpu" and e.get("name") in ZONES:
            per_zone.setdefault(e["name"], []).append(e["dur"] / 1000.0)

    # A zone that appears in under a third of the frames is a zone whose pass did not run
    # for most of the trace -- reporting a minimum over six samples as this tree's cost is
    # how a gate ends up defending a number that means nothing.
    return {k: min(v) for k, v in per_zone.items() if len(v) >= FRAMES // 3}


def main() -> int:
    ap = argparse.ArgumentParser(description=__doc__,
                                 formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("--config", default="release")
    ap.add_argument("--update", action="store_true", help="re-baseline, deliberately")
    args = ap.parse_args()

    now = measure(args.config)
    if not now:
        print("perfgate: the trace held none of the gated zones; not measured",
              file=sys.stderr)
        return 2

    if args.update or not BUDGET.exists():
        BUDGET.write_text(json.dumps(
            {"config": args.config, "frames": FRAMES, "arm": ARM, "marginPct": round((MARGIN - 1) * 100),
             "note": "Per-frame minimum in ms on this machine. scripts/perfgate.py --update rewrites it.",
             "zones": {k: round(v, 4) for k, v in sorted(now.items())}}, indent=2) + "\n")
        verb = "updated" if args.update else "created"
        print(f"perfgate: {verb} {BUDGET.name} -- " +
              "  ".join(f"{k} {v:.3f}" for k, v in sorted(now.items())))
        return 0

    budget = json.loads(BUDGET.read_text())
    if budget.get("config") != args.config:
        print(f"perfgate: budget is for {budget.get('config')}, measured {args.config}; "
              f"not comparable", file=sys.stderr)
        return 2

    over = []
    for zone in sorted(now):
        recorded = budget["zones"].get(zone)
        if recorded is None:
            continue
        limit = recorded * MARGIN
        flag = "OVER" if now[zone] > limit else "ok"
        print(f"  {zone:<10} {now[zone]:7.3f} ms   budget {recorded:.3f}  limit {limit:.3f}  {flag}")
        if now[zone] > limit:
            over.append((zone, now[zone], recorded))

    if not over:
        print("perfgate: inside budget")
        return 0

    print("\nperfgate: the frame got slower.", file=sys.stderr)
    for zone, got, recorded in over:
        print(f"  {zone} {got:.3f} ms against {recorded:.3f} "
              f"({(got / recorded - 1) * 100:+.1f}%)", file=sys.stderr)
    print("\nThis is a minimum over 300 frames, so it is not machine load. Either the change\n"
          "made the frame slower, or the cost is understood and intended -- in which case\n"
          "re-baseline on purpose with `scripts/perfgate.py --update` and say why in the\n"
          "commit. `git commit --no-verify` skips the gate and leaves no record that it did.",
          file=sys.stderr)
    return 1


if __name__ == "__main__":
    sys.exit(main())

File: tools/cli/Perfgate.cpp
#include "Bench.h"

#include "Process.h"
#include "Repo.h"
#include "Trace.h"

#include <rapidjson/document.h>
#include <rapidjson/istreamwrapper.h>

#include <algorithm>
#include <cmath>
#include <cstdio>
#include <fstream>
#include <map>
#include <sstream>

namespace tool {
namespace {

namespace fs = std::filesystem;

/// The arm. Fixed on purpose: a gate whose scene or camera moved with the tree would compare
/// two different pictures and call the difference a regression. `--locked` pins the clock,
/// `--headless` never maps a window so a commit hook cannot steal focus, and the camera is
/// written out rather than left to `frameBounds`, which follows the scene's bounds and would
/// shift the moment an asset did.
const char* const kArm[] = {"--locked",  "--audio-null", "--headless", "--msaa", "4",
                            "--camera",  "0.00,1.16,1.80,63.6,-12.6,7.87"};

constexpr int kFrames = 300;
const char* const kZones[] = {"Lighting", "Frame"};

/// 12%. Wide enough that ordinary machine noise does not fail a commit, narrow enough that a
/// real regression in either gated zone does.
constexpr double kMargin = 1.12;

/// Per-frame **minimum**, not median: the cheapest frame is the one least contaminated by
/// whatever else the machine was doing, which is what makes this a gate rather than a
/// weather report.
bool measure(Config config, std::map<std::string, double>& out) {
    const fs::path binary = buildDir(config) / executableName("demo");
    std::error_code ec;
    if (!fs::exists(binary, ec)) {
        std::fprintf(stderr, "perfgate: %s is not built. Run: substrate build-game demo %s\n",
                     binary.generic_string().c_str(), name(config));
        return false;
    }

    const fs::path dir = fs::temp_directory_path(ec) / "substrate-perfgate";
    fs::create_directories(dir, ec);
    const fs::path trace = dir / "trace.json";
    fs::remove(trace, ec);

    std::vector<std::string> argv{selfPath().string(), "run", name(config), "--"};
    for (const char* arg : kArm) argv.push_back(arg);
    argv.insert(argv.end(), {"--frames", std::to_string(kFrames), "--trace", trace.string()});

    RunOptions options;
    options.cwd = repoRoot();
    options.timeoutSeconds = 180;
    const RunResult result = run(argv, options);

    if (!fs::exists(trace, ec)) {
        std::fprintf(stderr, "perfgate: no trace written (exit %d); not measured\n",
                     result.exitCode);
        const size_t tail = result.err.size() > 2000 ? result.err.size() - 2000 : 0;
        std::fprintf(stderr, "%s\n", result.err.substr(tail).c_str());
        return false;
    }

    Trace parsed;
    std::string error;
    const bool read = readTrace(trace, parsed, error);
    fs::remove_all(dir, ec);
    if (!read) {
        std::fprintf(stderr, "perfgate: %s\n", error.c_str());
        return false;
    }

    for (const char* zone : kZones) {
        const auto found = parsed.gpu.find(zone);
        if (found == parsed.gpu.end()) continue;
        // A zone that appears in under a third of the frames is a zone whose pass did not
        // run for most of the trace -- reporting a minimum over six samples as this tree's
        // cost is how a gate ends up defending a number that means nothing.
        if (found->second.size() < static_cast<size_t>(kFrames / 3)) continue;
        out[zone] = minimum(found->second);
    }
    return true;
}

fs::path budgetPath() { return repoRoot() / "perf-budget.json"; }

void writeBudget(Config config, const std::map<std::string, double>& measured) {
    std::ostringstream json;
    json << "{\n";
    json << "  \"config\": \"" << name(config) << "\",\n";
    json << "  \"frames\": " << kFrames << ",\n";
    json << "  \"arm\": [";
    for (size_t i = 0; i < std::size(kArm); ++i) {
        json << (i ? ", " : "") << '"' << kArm[i] << '"';
    }
    json << "],\n";
    json << "  \"marginPct\": " << static_cast<int>(std::lround((kMargin - 1.0) * 100)) << ",\n";
    json << "  \"note\": \"Per-frame minimum in ms on this machine. `substrate perfgate "
            "--update` rewrites it.\",\n";
    json << "  \"zones\": {\n";
    size_t index = 0;
    for (const auto& [zone, ms] : measured) {
        char cell[64];
        std::snprintf(cell, sizeof(cell), "%.4f", ms);
        json << "    \"" << zone << "\": " << cell
             << (++index == measured.size() ? "\n" : ",\n");
    }
    json << "  }\n}\n";

    std::ofstream out(budgetPath());
    out << json.str();
}

} // namespace

int cmdPerfgate(const std::vector<std::string>& args) {
    Config config = Config::Release;
    bool update = false;

    for (size_t i = 0; i < args.size(); ++i) {
        if (args[i] == "--update") {
            update = true;
        } else if (args[i] == "--config" && i + 1 < args.size()) {
            const std::optional<Config> parsed = parseConfig(args[++i]);
            if (!parsed) {
                std::fprintf(stderr, "error: unknown config '%s' (want: %s)\n", args[i].c_str(),
                             configList().c_str());
                return 2;
            }
            config = *parsed;
        } else if (args[i] == "-h" || args[i] == "--help") {
            std::fputs("usage: substrate perfgate [--config <cfg>] [--update]\n", stderr);
            return 0;
        } else {
            std::fprintf(stderr, "error: unknown argument '%s'\n", args[i].c_str());
            return 2;
        }
    }

    std::map<std::string, double> now;
    if (!measure(config, now)) return 2;
    if (now.empty()) {
        std::fputs("perfgate: the trace held none of the gated zones; not measured\n", stderr);
        return 2;
    }

    std::error_code ec;
    if (update || !fs::exists(budgetPath(), ec)) {
        writeBudget(config, now);
        std::printf("perfgate: %s perf-budget.json --", update ? "updated" : "created");
        for (const auto& [zone, ms] : now) std::printf("  %s %.3f", zone.c_str(), ms);
        std::printf("\n");
        return 0;
    }

    std::ifstream in(budgetPath());
    rapidjson::IStreamWrapper stream(in);
    rapidjson::Document budget;
    budget.ParseStream(stream);
    if (budget.HasParseError() || !budget.IsObject()) {
        std::fputs("perfgate: perf-budget.json is malformed; not comparable\n", stderr);
        return 2;
    }

    const std::string recordedConfig =
        budget.HasMember("config") ? budget["config"].GetString() : "";
    if (recordedConfig != name(config)) {
        std::fprintf(stderr, "perfgate: budget is for %s, measured %s; not comparable\n",
                     recordedConfig.c_str(), name(config));
        return 2;
    }
    if (!budget.HasMember("zones") || !budget["zones"].IsObject()) {
        std::fputs("perfgate: perf-budget.json records no zones; not comparable\n", stderr);
        return 2;
    }

    struct Over {
        std::string zone;
        double got = 0.0;
        double recorded = 0.0;
    };
    std::vector<Over> over;

    for (const auto& [zone, ms] : now) {
        const rapidjson::Value& zones = budget["zones"];
        if (!zones.HasMember(zone.c_str())) continue;
        const double recorded = zones[zone.c_str()].GetDouble();
        const double limit = recorded * kMargin;
        std::printf("  %-10s %7.3f ms   budget %.3f  limit %.3f  %s\n", zone.c_str(), ms,
                    recorded, limit, ms > limit ? "OVER" : "ok");
        if (ms > limit) over.push_back({zone, ms, recorded});
    }

    if (over.empty()) {
        std::puts("perfgate: inside budget");
        return 0;
    }

    std::fputs("\nperfgate: the frame got slower.\n", stderr);
    for (const Over& entry : over) {
        std::fprintf(stderr, "  %s %.3f ms against %.3f (%+.1f%%)\n", entry.zone.c_str(),
                     entry.got, entry.recorded, (entry.got / entry.recorded - 1.0) * 100.0);
    }
    std::fputs("\nThis is a minimum over 300 frames, so it is not machine load. Either the "
               "change\nmade the frame slower, or the cost is understood and intended -- in "
               "which case\nre-baseline on purpose with `substrate perfgate --update` and say "
               "why in the\ncommit. `git commit --no-verify` skips the gate and leaves no "
               "record that it did.\n",
               stderr);
    return 1;
}

} // namespace tool
